File: LeetCode-Swift/LeetCode-Swift/二叉树/_94_二叉树的中序遍历.swift
//
//  _94_二叉树的中序遍历.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/12/9.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-tree-inorder-traversal/%20/

import Foundation

class _94_二叉树的中序遍历 {
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        var results = [Int]()
        var nodes = [TreeNode]()
        var p = root
        while p != nil || !nodes.isEmpty {
            while p != nil {
                nodes.append(p!)
                p = p?.left
            }
            if !nodes.isEmpty {
                results.append(nodes.last!.val)
                p = nodes.removeLast().right;
            }
        }
        return results
    }
}

/**
 输入: [1,null,2,3]
    1
     \
      2
     /
    3

 输出: [1,3,2]
 */
func test_94() {
    let root = generateTreeByLevel([1, nil, 2, 3])
    let solution = _94_二叉树的中序遍历()
    print(solution.inorderTraversal(root))
}

File: LeetCode-C++/LeetCode-C++/二叉树/_94_二叉树的中序遍历.cpp
//
//  _94_二叉树的中序遍历.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/12/9.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-tree-inorder-traversal/%20/

#include <stdio.h>
#include <vector>
#include <stack>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> results;
        stack<TreeNode*> st;
        TreeNode* p = root;
        while (p || !st.empty()) {
            while (p) {
                st.push(p);
                p = p->left;
            }
            if (!st.empty()) {
                results.push_back(st.top()->val);
                p = st.top()->right;
                st.pop();
            }
        }
        return results;
    }
};

/**
 输入: [1,null,2,3]
    1
     \
      2
     /
    3

 输出: [1,3,2]
 */
/**
int main(int argc, const char * argv[]) {
    TreeNode* root = generateTreeByLevel({1, INT_MAX, 2, 3});
    Solution* solution = new Solution();
    vector<int> result = solution->inorderTraversal(root);
    printVector(&result);
    return 0;
}
 */
